File: backend/app/api/routes/layer.py
"""API routes for layer computation visualization."""

from fastapi import APIRouter, HTTPException, status
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field

from app.models.schemas import TransformerConfig, TensorData
from app.core.layer_wrapper import (
    EncoderLayerWrapper,
    DecoderLayerWrapper,
    create_encoder_layer,
    create_decoder_layer,
    NANOTORCH_AVAILABLE,
)

router = APIRouter(prefix="/api/v1/layer", tags=["layer"])


# ============================================================================
# Request/Response Models
# ============================================================================

class LayerInput(BaseModel):
    """Input data for layer computation."""
    data: List[List[float]] = Field(..., description="Input tensor data (seq_len, d_model)")
    shape: List[int] = Field(..., description="Shape of the input tensor")
    dtype: str = Field("float32", description="Data type")


class EncoderLayerInput(BaseModel):
    """Input for encoder layer computation."""
    config: TransformerConfig = Field(..., description="Transformer configuration")
    input_data: LayerInput = Field(..., description="Layer input data")
    src_mask: Optional[List[List[float]]] = Field(None, description="Source attention mask")
    is_causal: bool = Field(False, description="Whether to use causal attention")


class DecoderLayerInput(BaseModel):
    """Input for decoder layer computation."""
    config: TransformerConfig = Field(..., description="Transformer configuration")
    input_data: LayerInput = Field(..., description="Decoder input data (target)")
    encoder_output: LayerInput = Field(..., description="Encoder output (memory)")
    tgt_mask: Optional[List[List[float]]] = Field(None, description="Target attention mask")
    memory_mask: Optional[List[List[float]]] = Field(None, description="Memory attention mask")


class LayerComputeOptions(BaseModel):
    """Options for layer computation."""
    return_attention_weights: bool = Field(True, description="Return attention weights")
    return_parameters: bool = Field(False, description="Return layer parameters")
    batch_size: int = Field(1, ge=1, le=8, description="Batch size for computation")


class SublayerComputation(BaseModel):
    """Results from a single sublayer computation."""
    norm_input: Optional[TensorData] = Field(None, description="Input to layer norm")
    norm_output: Optional[TensorData] = Field(None, description="Output from layer norm")
    attention: Optional[Dict[str, Any]] = Field(None, description="Attention computation results")
    linear1_output: Optional[TensorData] = Field(None, description="Output from first linear layer")
    activation_output: Optional[TensorData] = Field(None, description="Output after activation")
    linear2_output: Optional[TensorData] = Field(None, description="Output from second linear layer")
    dropout_output: Optional[TensorData] = Field(None, description="Output after dropout")
    residual_output: TensorData = Field(..., description="Output after residual connection")


class EncoderLayerResult(BaseModel):
    """Results from encoder layer computation."""
    success: bool = Field(..., description="Whether computation was successful")
    input: TensorData = Field(..., description="Layer input")
    config: Dict[str, Any] = Field(..., description="Layer configuration")
    sublayer1: SublayerComputation = Field(..., description="Self-attention sublayer results")
    sublayer2: SublayerComputation = Field(..., description="Feed-forward sublayer results")
    output: TensorData = Field(..., description="Layer output")
    error: Optional[str] = Field(None, description="Error message if failed")


class DecoderLayerResult(BaseModel):
    """Results from decoder layer computation."""
    success: bool = Field(..., description="Whether computation was successful")
    input: TensorData = Field(..., description="Layer input (target)")
    encoder_output: TensorData = Field(..., description="Encoder output (memory)")
    config: Dict[str, Any] = Field(..., description="Layer configuration")
    sublayer1: SublayerComputation = Field(..., description="Masked self-attention sublayer results")
    sublayer2: SublayerComputation = Field(..., description="Cross-attention sublayer results")
    sublayer3: SublayerComputation = Field(..., description="Feed-forward sublayer results")
    output: TensorData = Field(..., description="Layer output")
    error: Optional[str] = Field(None, description="Error message if failed")


class LayerStatistics(BaseModel):
    """Statistics about layer computation."""
    num_parameters: int = Field(..., description="Total number of parameters")
    flops: int = Field(..., description="Approximate FLOPs for forward pass")
    memory_mb: float = Field(..., description="Approximate memory usage in MB")
    sublayer_breakdown: Dict[str, int] = Field(..., description="Parameters per sublayer")


# ============================================================================
# Helper Functions
# ============================================================================

def calculate_layer_stats(config: TransformerConfig, layer_type: str = "encoder") -> LayerStatistics:
    """Calculate statistics for a layer.

    Args:
        config: Transformer configuration
        layer_type: Type of layer ("encoder" or "decoder")

    Returns:
        Layer statistics
    """
    d_model = config.d_model
    nhead = config.nhead
    dim_feedforward = config.dim_feedforward

    # Calculate parameters for each component
    # Self-attention: Q, K, V projections + output projection
    attn_params = 4 * d_model * d_model  # 4 weight matrices
    attn_params += 4 * d_model  # 4 bias vectors

    # Feed-forward: 2 linear layers
    ffn_params = d_model * dim_feedforward + dim_feedforward  # Linear1
    ffn_params += dim_feedforward * d_model + d_model  # Linear2

    # Layer norms (2 for encoder, 3 for decoder)
    num_norms = 3 if layer_type == "decoder" else 2
    norm_params = num_norms * (2 * d_model)  # gamma and beta for each norm

    # Total parameters
    total_params = attn_params + ffn_params + norm_params

    # Cross-attention for decoder (same as self-attention)
    if layer_type == "decoder":
        total_params += attn_params

    # Estimate FLOPs (very rough approximation)
    seq_len = 64  # Assume sequence length
    flops = 2 * seq_len * seq_len * d_model  # Self-attention
    flops += 2 * seq_len * d_model * dim_feedforward  # FFN
    flops *= 2  # Multiply by 2 for decoder (cross-attn)

    # Estimate memory
    memory_mb = (total_params * 4) / (1024 * 1024)  # float32 = 4 bytes

    return LayerStatistics(
        num_parameters=total_params,
        flops=flops,
        memory_mb=memory_mb,
        sublayer_breakdown={
            "self_attention": attn_params,
            "feed_forward": ffn_params,
            "layer_norm": norm_params,
        }
    )


# ============================================================================
# API Endpoints
# ============================================================================

@router.post("/encoder/compute", response_model=EncoderLayerResult)
async def compute_encoder_layer(request: EncoderLayerInput):
    """Compute encoder layer with intermediate results.

    Args:
        request: Encoder layer computation request

    Returns:
        Detailed computation results with all intermediate tensors
    """
    if not NANOTORCH_AVAILABLE:
        return EncoderLayerResult(
            success=False,
            input=TensorData(shape=[0], data=[], dtype="float32"),
            config={},
            sublayer1=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            sublayer2=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            output=TensorData(shape=[0], data=[], dtype="float32"),
            error="nanotorch is not available"
        )

    try:
        # Create encoder layer wrapper
        layer = create_encoder_layer(request.config.dict())

        # Prepare input
        input_data = request.input_data.data

        # Run forward pass with intermediates
        results = layer.forward_with_intermediates(
            src=input_data,
            src_mask=request.src_mask,
            is_causal=request.is_causal,
        )

        # Add success flag
        results['success'] = True
        results['error'] = None

        return results

    except Exception as e:
        return EncoderLayerResult(
            success=False,
            input=TensorData(
                shape=request.input_data.shape,
                data=request.input_data.data,
                dtype=request.input_data.dtype
            ),
            config=request.config.dict(),
            sublayer1=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            sublayer2=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            output=TensorData(shape=[0], data=[], dtype="float32"),
            error=f"Computation failed: {str(e)}"
        )


@router.post("/decoder/compute", response_model=DecoderLayerResult)
async def compute_decoder_layer(request: DecoderLayerInput):
    """Compute decoder layer with intermediate results.

    Args:
        request: Decoder layer computation request

    Returns:
        Detailed computation results with all intermediate tensors
    """
    if not NANOTORCH_AVAILABLE:
        return DecoderLayerResult(
            success=False,
            input=TensorData(shape=[0], data=[], dtype="float32"),
            encoder_output=TensorData(shape=[0], data=[], dtype="float32"),
            config={},
            sublayer1=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            sublayer2=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            sublayer3=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            output=TensorData(shape=[0], data=[], dtype="float32"),
            error="nanotorch is not available"
        )

    try:
        # Create decoder layer wrapper
        layer = create_decoder_layer(request.config.dict())

        # Prepare inputs
        tgt_data = request.input_data.data
        memory_data = request.encoder_output.data

        # Run forward pass with intermediates
        results = layer.forward_with_intermediates(
            tgt=tgt_data,
            memory=memory_data,
            tgt_mask=request.tgt_mask,
            memory_mask=request.memory_mask,
        )

        # Add success flag
        results['success'] = True
        results['error'] = None

        return results

    except Exception as e:
        return DecoderLayerResult(
            success=False,
            input=TensorData(
                shape=request.input_data.shape,
                data=request.input_data.data,
                dtype=request.input_data.dtype
            ),
            encoder_output=TensorData(
                shape=request.encoder_output.shape,
                data=request.encoder_output.data,
                dtype=request.encoder_output.dtype
            ),
            config=request.config.dict(),
            sublayer1=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            sublayer2=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            sublayer3=SublayerComputation(
                residual_output=TensorData(shape=[0], data=[], dtype="float32")
            ),
            output=TensorData(shape=[0], data=[], dtype="float32"),
            error=f"Computation failed: {str(e)}"
        )


@router.post("/encoder/statistics", response_model=LayerStatistics)
async def get_encoder_statistics(config: TransformerConfig):
    """Get statistics for an encoder layer.

    Args:
        config: Transformer configuration

    Returns:
        Layer statistics including parameters, FLOPs, memory
    """
    return calculate_layer_stats(config, "encoder")


@router.post("/decoder/statistics", response_model=LayerStatistics)
async def get_decoder_statistics(config: TransformerConfig):
    """Get statistics for a decoder layer.

    Args:
        config: Transformer configuration

    Returns:
        Layer statistics including parameters, FLOPs, memory
    """
    stats = calculate_layer_stats(config, "decoder")
    # Add cross-attention parameters for decoder
    d_model = config.d_model
    cross_attn_params = 4 * d_model * d_model + 4 * d_model
    stats.num_parameters += cross_attn_params
    stats.sublayer_breakdown["cross_attention"] = cross_attn_params
    return stats


@router.get("/types")
async def get_layer_types():
    """Get available layer types and their descriptions.

    Returns:
        List of layer types with descriptions
    """
    return {
        "success": True,
        "layer_types": [
            {
                "type": "encoder",
                "name": "Encoder Layer",
                "description": "Transformer encoder layer with self-attention and feed-forward network",
                "sublayers": ["Self-Attention", "Feed-Forward Network"],
                "architecture": "Post-Norm or Pre-Norm"
            },
            {
                "type": "decoder",
                "name": "Decoder Layer",
                "description": "Transformer decoder layer with masked self-attention, cross-attention, and feed-forward network",
                "sublayers": ["Masked Self-Attention", "Cross-Attention", "Feed-Forward Network"],
                "architecture": "Post-Norm or Pre-Norm"
            }
        ]
    }

File: frontend/src/types/layer.ts
/**
 * TypeScript types for layer computation visualization.
 *
 * This module defines all types related to Transformer encoder and decoder layer
 * computation visualization, including intermediate results, step-by-step processing,
 * and layer statistics.
 */

/**
 * Supported layer types.
 */
export type LayerType = 'encoder' | 'decoder';

/**
 * Sublayer types within a Transformer layer.
 */
export type SublayerType = 'self_attention' | 'cross_attention' | 'feedforward';

/**
 * Tensor data representation.
 */
export interface TensorData {
  /** Shape of the tensor (e.g., [batch_size, seq_len, d_model]) */
  shape: number[];
  /** Tensor data (flattened or nested depending on dimensionality) */
  data: number[] | number[][] | number[][][] | number[][][][];
  /** Data type (e.g., 'float32') */
  dtype: string;
  /** Optional tensor name */
  name?: string;
}

/**
 * Attention computation results.
 */
export interface AttentionComputation {
  /** Output from attention computation */
  output: TensorData;
  /** Query tensor (optional, if returned) */
  queries?: TensorData;
  /** Key tensor (optional, if returned) */
  keys?: TensorData;
  /** Value tensor (optional, if returned) */
  values?: TensorData;
  /** Attention scores before softmax */
  scores?: TensorData;
  /** Attention weights (after softmax) */
  weights?: TensorData;
  /** Scale factor used in attention */
  scale?: number;
}

/**
 * Results from a single sublayer computation.
 */
export interface SublayerComputation {
  /** Input to layer normalization */
  norm_input?: TensorData;
  /** Output from layer normalization */
  norm_output?: TensorData;
  /** Attention computation results (for attention sublayers) */
  attention?: AttentionComputation;
  /** Output from first linear layer (FFN) */
  linear1_output?: TensorData;
  /** Output after activation function (FFN) */
  activation_output?: TensorData;
  /** Output from second linear layer (FFN) */
  linear2_output?: TensorData;
  /** Output after dropout (first dropout in sublayer) */
  dropout_output?: TensorData;
  /** Additional dropout output (for FFN with two dropouts) */
  dropout1_output?: TensorData;
  /** Additional dropout output (for FFN with two dropouts) */
  dropout2_output?: TensorData;
  /** Output after residual connection (x + Sublayer(x)) */
  residual_output: TensorData;
}

/**
 * Complete encoder layer computation results.
 */
export interface EncoderLayerResult {
  /** Whether the computation was successful */
  success: boolean;
  /** Layer input tensor */
  input: TensorData;
  /** Layer configuration */
  config: LayerConfig;
  /** Self-attention sublayer results */
  sublayer1: SublayerComputation;
  /** Feed-forward sublayer results */
  sublayer2: SublayerComputation;
  /** Layer output tensor */
  output: TensorData;
  /** Error message if computation failed */
  error?: string;
}

/**
 * Complete decoder layer computation results.
 */
export interface DecoderLayerResult {
  /** Whether the computation was successful */
  success: boolean;
  /** Layer input tensor (decoder input/target) */
  input: TensorData;
  /** Encoder output (memory) for cross-attention */
  encoder_output: TensorData;
  /** Layer configuration */
  config: LayerConfig;
  /** Masked self-attention sublayer results */
  sublayer1: SublayerComputation;
  /** Cross-attention sublayer results */
  sublayer2: SublayerComputation;
  /** Feed-forward sublayer results */
  sublayer3: SublayerComputation;
  /** Layer output tensor */
  output: TensorData;
  /** Error message if computation failed */
  error?: string;
}

/**
 * Layer configuration parameters.
 */
export interface LayerConfig {
  /** Model dimension */
  d_model: number;
  /** Number of attention heads */
  nhead: number;
  /** Feed-forward network dimension */
  dim_feedforward: number;
  /** Dropout rate */
  dropout: number;
  /** Activation function ('relu' or 'gelu') */
  activation: string;
  /** Whether to use pre-layer normalization */
  norm_first: boolean;
  /** Layer normalization epsilon */
  layer_norm_eps: number;
  /** Whether batch dimension comes first */
  batch_first: boolean;
}

/**
 * Input data for layer computation.
 */
export interface LayerInput {
  /** Input tensor data */
  data: number[][];
  /** Shape of the input tensor */
  shape: number[];
  /** Data type */
  dtype: string;
}

/**
 * Request for encoder layer computation.
 */
export interface EncoderLayerRequest {
  /** Transformer configuration */
  config: LayerConfig;
  /** Layer input data */
  input_data: LayerInput;
  /** Source attention mask (optional) */
  src_mask?: number[][];
  /** Whether to use causal attention */
  is_causal: boolean;
}

/**
 * Request for decoder layer computation.
 */
export interface DecoderLayerRequest {
  /** Transformer configuration */
  config: LayerConfig;
  /** Decoder input data (target) */
  input_data: LayerInput;
  /** Encoder output (memory) */
  encoder_output: LayerInput;
  /** Target attention mask (optional) */
  tgt_mask?: number[][];
  /** Memory attention mask (optional) */
  memory_mask?: number[][];
}

/**
 * Options for layer computation.
 */
export interface LayerComputeOptions {
  /** Whether to return attention weights */
  return_attention_weights?: boolean;
  /** Whether to return layer parameters */
  return_parameters?: boolean;
  /** Batch size for computation */
  batch_size?: number;
}

/**
 * Statistics about layer computation.
 */
export interface LayerStatistics {
  /** Total number of parameters */
  num_parameters: number;
  /** Approximate FLOPs for forward pass */
  flops: number;
  /** Approximate memory usage in MB */
  memory_mb: number;
  /** Parameters per sublayer/component */
  sublayer_breakdown: {
    self_attention?: number;
    cross_attention?: number;
    feed_forward?: number;
    layer_norm?: number;
    [key: string]: number | undefined;
  };
}

/**
 * Response from layer statistics endpoint.
 */
export interface LayerStatisticsResponse {
  /** Total number of parameters */
  num_parameters: number;
  /** Approximate FLOPs for forward pass */
  flops: number;
  /** Approximate memory usage in MB */
  memory_mb: number;
  /** Parameters per sublayer/component */
  sublayer_breakdown: {
    self_attention?: number;
    cross_attention?: number;
    feed_forward?: number;
    layer_norm?: number;
    [key: string]: number | undefined;
  };
}

/**
 * Layer computation step for visualization navigation.
 */
export interface LayerStep {
  /** Step identifier */
  id: string;
  /** Step name (e.g., 'input', 'norm1', 'attention', etc.) */
  name: string;
  /** Step display title */
  title: string;
  /** Step description */
  description: string;
  /** Step index in the sequence */
  index: number;
  /** Whether this step is currently active */
  isActive: boolean;
  /** Whether this step has been completed */
  isCompleted: boolean;
  /** Associated sublayer (0, 1, or 2) */
  sublayer?: number;
}

/**
 * Computation stage within a sublayer.
 */
export type ComputationStage =
  | 'input'
  | 'norm'
  | 'attention'
  | 'linear1'
  | 'activation'
  | 'linear2'
  | 'dropout'
  | 'residual'
  | 'output';

/**
 * Data flow node for visualization.
 */
export interface DataFlowNode {
  /** Node identifier */
  id: string;
  /** Node label */
  label: string;
  /** Node type */
  type: 'operation' | 'data' | 'output';
  /** X position */
  x: number;
  /** Y position */
  y: number;
  /** Associated tensor shape */
  shape?: number[];
  /** Associated computation stage */
  stage?: ComputationStage;
}

/**
 * Data flow edge for visualization.
 */
export interface DataFlowEdge {
  /** Edge identifier */
  id: string;
  /** Source node ID */
  source: string;
  /** Target node ID */
  target: string;
  /** Edge label */
  label?: string;
  /** Whether this is a residual connection */
  isResidual?: boolean;
}

/**
 * Complete data flow diagram.
 */
export interface DataFlowDiagram {
  /** All nodes in the flow */
  nodes: DataFlowNode[];
  /** All edges in the flow */
  edges: DataFlowEdge[];
  /** Currently highlighted step */
  highlightStep?: string;
}

/**
 * Tensor statistics for visualization.
 */
export interface TensorStatistics {
  /** Minimum value */
  min: number;
  /** Maximum value */
  max: number;
  /** Mean value */
  mean: number;
  /** Standard deviation */
  std: number;
  /** Median value */
  median: number;
  /** 25th percentile */
  q25: number;
  /** 75th percentile */
  q75: number;
}

/**
 * Comparison between two tensors.
 */
export interface TensorComparison {
  /** Before tensor */
  before: TensorData;
  /** After tensor */
  after: TensorData;
  /** Operation performed */
  operation: string;
  /** Statistics before */
  beforeStats: TensorStatistics;
  /** Statistics after */
  afterStats: TensorStatistics;
  /** Difference statistics */
  differenceStats?: TensorStatistics;
}

/**
 * Layer visualization state.
 */
export interface LayerVisualizationState {
  /** Currently selected layer type */
  layerType: LayerType;
  /** Currently selected layer index */
  layerIndex: number;
  /** Current step in the visualization */
  currentStep: number;
  /** Whether to show tensor values */
  showValues: boolean;
  /** Whether to show tensor shapes */
  showShapes: boolean;
  /** Whether to show statistics */
  showStatistics: boolean;
  /** Whether to animate transitions */
  animateTransitions: boolean;
  /** Computation result */
  computationResult: EncoderLayerResult | DecoderLayerResult | null;
  /** Whether computation is in progress */
  isComputing: boolean;
  /** Computation error */
  computationError: string | null;
  /** Playback state */
  isPlaying: boolean;
  /** Playback speed (ms per step) */
  playbackSpeed: number;
}

/**
 * Layer information.
 */
export interface LayerInfo {
  /** Layer type */
  type: LayerType;
  /** Layer index */
  index: number;
  /** Layer configuration */
  config: LayerConfig;
  /** Number of sublayers */
  numSublayers: number;
  /** Sublayer names */
  sublayerNames: string[];
}

/**
 * Layer type information.
 */
export interface LayerTypeInfo {
  /** Layer type */
  type: LayerType;
  /** Display name */
  name: string;
  /** Description */
  description: string;
  /** Sublayers in this layer type */
  sublayers: string[];
  /** Architecture type (post-norm or pre-norm) */
  architecture: string;
}

/**
 * Response from layer types endpoint.
 */
export interface LayerTypesResponse {
  /** Whether request was successful */
  success: boolean;
  /** Available layer types */
  layer_types: LayerTypeInfo[];
}

/**
 * Progress information for layer computation.
 */
export interface LayerComputationProgress {
  /** Current step being computed */
  currentStep: number;
  /** Total number of steps */
  totalSteps: number;
  /** Current sublayer being computed */
  currentSublayer: number;
  /** Percentage complete */
  percentage: number;
}

/**
 * Error information for layer computation.
 */
export interface LayerComputationError {
  /** Error code */
  code: string;
  /** Error message */
  message: string;
  /** Stack trace (if available) */
  stackTrace?: string;
  /** Step where error occurred */
  errorStep?: string;
}
